File: src/NodeParser/AnnotatedNodeParser.ts
import ts from "typescript";
import { AnnotationsReader } from "../AnnotationsReader.js";
import { ExtendedAnnotationsReader } from "../AnnotationsReader/ExtendedAnnotationsReader.js";
import { Context } from "../NodeParser.js";
import { SubNodeParser } from "../SubNodeParser.js";
import { AnnotatedType } from "../Type/AnnotatedType.js";
import { BaseType } from "../Type/BaseType.js";
import { ReferenceType } from "../Type/ReferenceType.js";
import { removeUndefined } from "../Utils/removeUndefined.js";
import { DefinitionType } from "../Type/DefinitionType.js";
import { UnionType } from "../Type/UnionType.js";
import { AnyType } from "../Type/AnyType.js";

export class AnnotatedNodeParser implements SubNodeParser {
    public constructor(
        protected childNodeParser: SubNodeParser,
        protected annotationsReader: AnnotationsReader,
    ) {}

    public supportsNode(node: ts.Node): boolean {
        return this.childNodeParser.supportsNode(node);
    }

    public createType(node: ts.Node, context: Context, reference?: ReferenceType): BaseType | undefined {
        const annotatedNode = this.getAnnotatedNode(node);
        let annotations = this.annotationsReader.getAnnotations(annotatedNode);
        const nullable = this.getNullable(annotatedNode);

        // Short-circuit parsing the underlying type if an explicit ref annotation was passed.
        if (annotations && "$ref" in annotations) {
            return new AnnotatedType(new AnyType(), annotations, nullable);
        }

        const baseType = this.childNodeParser.createType(node, context, reference);

        if (!baseType) {
            return undefined;
        }

        // Don't return annotations for lib types such as Exclude.
        // Sourceless nodes may not have a fileName, just ignore them.
        if (node.getSourceFile()?.fileName.match(/[/\\]typescript[/\\]lib[/\\]lib\.[^/\\]+\.d\.ts$/i)) {
            let specialCase = false;

            // Special case for Exclude<T, U>: use the annotation of T.
            if (
                node.kind === ts.SyntaxKind.TypeAliasDeclaration &&
                (node as ts.TypeAliasDeclaration).name.text === "Exclude"
            ) {
                let t = context.getArgument("T");

                // Handle optional properties.
                if (t instanceof UnionType) {
                    t = removeUndefined(t).newType;
                }

                if (t instanceof DefinitionType) {
                    t = t.getType();
                }

                if (t instanceof AnnotatedType) {
                    annotations = t.getAnnotations();
                    specialCase = true;
                }
            }

            if (!specialCase) {
                return baseType;
            }
        }

        return !annotations && !nullable ? baseType : new AnnotatedType(baseType, annotations || {}, nullable);
    }

    protected getNullable(annotatedNode: ts.Node) {
        return this.annotationsReader instanceof ExtendedAnnotationsReader
            ? this.annotationsReader.isNullable(annotatedNode)
            : false;
    }

    protected getAnnotatedNode(node: ts.Node): ts.Node {
        if (!node.parent) {
            return node;
        } else if (node.parent.kind === ts.SyntaxKind.PropertySignature) {
            return node.parent;
        } else if (node.parent.kind === ts.SyntaxKind.PropertyDeclaration) {
            return node.parent;
        } else if (node.parent.kind === ts.SyntaxKind.IndexSignature) {
            return node.parent;
        } else if (node.parent.kind === ts.SyntaxKind.Parameter) {
            return node.parent;
        } else {
            return node;
        }
    }
}

File: internal/parser/annotated_node_parser.go
package parser

import (
	"regexp"

	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// AnnotatedNodeParser decorates a child parser with JSDoc annotations
// (src/NodeParser/AnnotatedNodeParser.ts).
type AnnotatedNodeParser struct {
	childNodeParser   SubNodeParser
	annotationsReader AnnotationsReader
}

func NewAnnotatedNodeParser(childNodeParser SubNodeParser, annotationsReader AnnotationsReader) *AnnotatedNodeParser {
	return &AnnotatedNodeParser{childNodeParser: childNodeParser, annotationsReader: annotationsReader}
}

func (p *AnnotatedNodeParser) SupportsNode(node *ast.Node) bool {
	return p.childNodeParser.SupportsNode(node)
}

// libFilePattern matches TypeScript lib declaration files. The upstream
// pattern matches paths like .../typescript/lib/lib.es5.d.ts; typescript-go
// serves its bundled libs as bundled:///libs/lib.es5.d.ts, so those are
// matched as well.
var libFilePattern = regexp.MustCompile(`(?i)[/\\]typescript[/\\]lib[/\\]lib\.[^/\\]+\.d\.ts$|^bundled:///libs/lib\.[^/]+\.d\.ts$`)

func (p *AnnotatedNodeParser) CreateType(node *ast.Node, ctx *Context, reference *types.ReferenceType) types.Type {
	annotatedNode := getAnnotatedNode(node)
	annotations := p.annotationsReader.GetAnnotations(annotatedNode)
	nullable := p.getNullable(annotatedNode)

	// Short-circuit parsing the underlying type if an explicit ref annotation was passed.
	if _, hasRef := annotations["$ref"]; hasRef {
		return &types.AnnotatedType{Type: &types.AnyType{}, Annotations: annotations, Nullable: nullable}
	}

	baseType := p.childNodeParser.CreateType(node, ctx, reference)

	// Don't return annotations for lib types such as Exclude.
	// Sourceless nodes may not have a fileName, just ignore them.
	if source := ast.GetSourceFileOfNode(node); source != nil && libFilePattern.MatchString(source.FileName()) {
		specialCase := false

		// Special case for Exclude<T, U>: use the annotation of T.
		if node.Kind == ast.KindTypeAliasDeclaration && node.Name() != nil && node.Name().Text() == "Exclude" {
			t := ctx.GetArgument("T")

			// Handle optional properties.
			if union, ok := t.(*types.UnionType); ok {
				_, t = types.RemoveUndefined(union)
			}

			if def, ok := t.(*types.DefinitionType); ok {
				t = def.Type
			}

			if annotated, ok := t.(*types.AnnotatedType); ok {
				annotations = annotated.Annotations
				if annotations == nil {
					annotations = types.Annotations{}
				}
				specialCase = true
			}
		}

		if !specialCase {
			return baseType
		}
	}

	if annotations == nil && !nullable {
		return baseType
	}
	if annotations == nil {
		annotations = types.Annotations{}
	}
	return &types.AnnotatedType{Type: baseType, Annotations: annotations, Nullable: nullable}
}

func (p *AnnotatedNodeParser) getNullable(annotatedNode *ast.Node) bool {
	if extended, ok := p.annotationsReader.(*ExtendedAnnotationsReader); ok {
		return extended.IsNullable(annotatedNode)
	}
	return false
}

// getAnnotatedNode hops one level to the parent when the parent is the
// declaration carrying the JSDoc (property, index signature, or parameter).
func getAnnotatedNode(node *ast.Node) *ast.Node {
	if node.Parent == nil {
		return node
	}
	switch node.Parent.Kind {
	case ast.KindPropertySignature, ast.KindPropertyDeclaration, ast.KindIndexSignature, ast.KindParameter:
		return node.Parent
	}
	return node
}
